overlord/snapstate/target.go

```diff
@@ -0,0 +1,569 @@
+		invalidSets, err := vsets.CheckPresenceInvalid(naming.Snap(t.InstanceName))
+		if err != nil {
+			if _, ok := err.(*snapasserts.PresenceConstraintError); !ok {
+				return nil, err
+			} // else presence is optional or required, carry on
+		}
+
+		if len(invalidSets) > 0 {
+			return nil, fmt.Errorf(
+				"cannot install snap %q due to enforcing rules of validation set %s",
+				t.InstanceName, snapasserts.ValidationSetKeySlice(invalidSets).CommaSeparated(),
+			)
+		}
+
+		requiredSets, requiredRev, err := vsets.CheckPresenceRequired(naming.Snap(t.InstanceName))
+		if err != nil {
+			return nil, err
+		}
+
+		// make sure that the caller-requested revision matches the revision
+		// required by the enforced validation sets
+		if requiredRev.Set() && t.RevOpts.Revision.Set() && requiredRev != t.RevOpts.Revision {
+			return nil, fmt.Errorf(
+				"cannot install snap %q at requested revision %s without --ignore-validation, revision %s required by validation sets: %s",
+				t.InstanceName, t.RevOpts.Revision, requiredRev, snapasserts.ValidationSetKeySlice(requiredSets).CommaSeparated(),
+			)
+		}
+
+		// TODO: handle validation sets and components here
+
+		action.ValidationSets = requiredSets
+
+		if requiredRev.Set() {
+			// make sure that we use the revision required by the enforced
+			// validation sets
+			action.Revision = requiredRev
+
+			// we ignore the cohort if a validation set requires that the
+			// snap is pinned to a specific revision
+			action.CohortKey = ""
+		}
+	}
+
+	// clear out the channel if we're requesting a specific revision, which
+	// could be because the user requested a specific revision or because a
+	// validation set requires it
+	if action.Revision.Set() {
+		action.Channel = ""
+	}
+
+	return action, nil
+}
+
+func (s *storeInstallGoal) validateAndPrune(installedSnaps map[string]*SnapState) error {
+	for name, t := range s.snaps {
+		if err := snap.ValidateInstanceName(name); err != nil {
+			return fmt.Errorf("invalid instance name: %v", err)
+		}
+
+		if err := validateRevisionOpts(t.RevOpts); err != nil {
+			return fmt.Errorf("invalid revision options for snap %q: %w", name, err)
+		}
+
+		snapst, ok := installedSnaps[name]
+		if ok && snapst.IsInstalled() {
+			if !t.SkipIfPresent {
+				return &snap.AlreadyInstalledError{Snap: name}
+			}
+			delete(s.snaps, name)
+		}
+	}
+	return nil
+}
+
+// InstallOne is a wrapper for InstallTarget that ensures that the given Target
+// installs exactly one snap. If the Target does not install exactly one snap,
+// an error is returned.
+func InstallOne(ctx context.Context, st *state.State, goal InstallGoal, opts Options) (*snap.Info, *state.TaskSet, error) {
+	infos, tasksets, err := InstallWithGoal(ctx, st, goal, opts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(infos) != 1 || len(tasksets) != 1 {
+		return nil, nil, errors.New("internal error: expected exactly one snap and taskset")
+	}
```

Comment:
this is a bit strange because one would expect that we could check whether goal return one target or not earlier

Reply:
Yes, this is a bit annoying. I was trying to avoid a flag being passed into `InstallWithGoal` to check that. One option is some implementation of `InstallGoal` that wraps another `InstallGoal` that makes sure that we return a singular `Target`.

Reply:
Might be a bit too complex though for something that really could just be an `if`.

Reply:
This can also be handled with the RequireOneSnap option introduced in 938cd26c6694412ea13e8bb443e0c164b3aa6e1f.